[DDP] Add host-side time to CUDATimer

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62770
* #62751
* #62748

Adding timing of forward, backward comp, backward comm, etc will help
detect desynchronization issues.

Differential Revision: [D30115585](https://our.internmc.facebook.com/intern/diff/D30115585/)

rohan-varma committed:
[DDP] Add host-side time to CUDATimer

Adding timing of forward, backward comp, backward comm, etc will help
detect desynchronization issues.

Differential Revision: [D30115585](https://our.internmc.facebook.com/intern/diff/D30115585/)

[ghstack-poisoned]

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -27,11 +27,27 @@ constexpr int kDefaultFirstBucketBytes = int(1024 * 1024);
 class TORCH_API Timer {
+  private:
+    // The timestamp of forward call start time in each iteration.
+    int64_t forward_start_time = kUnsetTime;
```
[Optional] Actually it is a mixed coding style. Personally prefer also using c10::optional<int64_t> instead of int64_t here, so no need create a `kUnsetTime` and have separate processing (that maps `kUnsetTime` to `nullopt`).

I understand this is not new code brought in by this PR. No need to do this code improvement in this PR.

rohan-varma replied:
Filed https://github.com/pytorch/pytorch/issues/62862, it can be a good ramp up task

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -41,14 +57,45 @@ class TORCH_API Timer {
+    }
+  }
+
+  // Return host-side time member variable corresponding to the given event.
```
Since it's specific to "host-side", will it make more sense to move it back to `CPUTime` class? Or do you want to update the comment?

rohan-varma replied:
See above comment regarding this

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -41,14 +57,45 @@ class TORCH_API Timer {
 
   virtual ~Timer() = default;
+
+  // Return host-side timestamp, or nullopt if it has not yet been recorded.
```
Since it's specific to "host-side", will it make more sense to move it back to CPUTime class? Or do you want to update the comment?

rohan-varma replied:
In the logger, this is also accessed for GPU timing events so we can get the host-side time, so I think it makes sense to keep it in the parent class?

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -41,14 +57,45 @@ class TORCH_API Timer {
+  // Record the current event, i.e., mark it as having occurred now. Default
+  // CPU implementation.
+  virtual void record(Event event) {
+    getTime(event) = current_time_in_nanos();
```
Nit: Better to rename it as `getTimeRef`, since this method is not a common read-only getter, but returns a mutable reference.

SciPioneer commented on torch/csrc/distributed/c10d/logger.cpp:
```diff
@@ -192,6 +192,17 @@ void Logger::set_construction_data_and_log(
+    Timer::Event event) {
+  auto timestamp = timer.getTimestamp(event);
+  if (timestamp != c10::nullopt) {
+    // TODO: should we set this as human-readable time instead of unixtime?
```
It seems that what are mainly used are the results returned by `measureDifference`, so probably fine.

rohan-varma committed:
Update on "[DDP] Add host-side time to CUDATimer"


Adding timing of forward, backward comp, backward comm, etc will help
detect desynchronization issues.

Differential Revision: [D30115585](https://our.internmc.facebook.com/intern/diff/D30115585/)

[ghstack-poisoned]
